lib/include/platform/iot_platform_types_afr.h

```diff
@@ -0,0 +1,92 @@
+/*
+ * Copyright (C) 2019 Amazon.com, Inc. or its affiliates.  All Rights Reserved.
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy of
+ * this software and associated documentation files (the "Software"), to deal in
+ * the Software without restriction, including without limitation the rights to
+ * use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of
+ * the Software, and to permit persons to whom the Software is furnished to do so,
+ * subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in all
+ * copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS
+ * FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR
+ * COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER
+ * IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
+ * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+ */
+
+/**
+ * @file iot_platform_types_posix.h
+ * @brief Definitions of platform layer types on POSIX systems.
+ */
+
+#ifndef _IOT_PLATFORM_TYPES_POSIX_H_
+#define _IOT_PLATFORM_TYPES_POSIX_H_
+
+#include "timers.h"
+
+
+typedef struct _iot_mutex_internal
+{
+    StaticSemaphore_t xMutex;           /**< FreeRTOS mutex. */
+    bool recursive;                           /**< Type; used for indicating if this is reentrant or normal. */
+} _iot_mutex_internal_t;
+
+/**
+ * @brief The native mutex type on AFR systems.
+ */
+typedef _iot_mutex_internal_t _IotSystemMutex_t;
+
+
+typedef struct _iot_sem_internal
+{
+    StaticSemaphore_t xSemaphore;       /**< FreeRTOS semaphore. */
+} _iot_sem_internal_t;
+
+/**
+ * @brief The native semaphore type on AFR systems.
+ */
+typedef _iot_sem_internal_t _IotSystemSemaphore_t;
```

Comment:
These types seem redundant, they are used in iot_platform_types.h to tie the internal semaphore struct to the specific underlying implementation via typdef proxy. I think it would be cleaner if the contract was implicit and implementors created the internal type as e.g. _IotSystemSemaphore_t directly and remove the proxy from platform_types? Wherever the type was used (IotMutex_t e.g.) it was always required that the implementation specific platform types was included in that file as well via the config file.

Opinions about the design here please

Reply:
Your proposal makes sense to me, as it does the current implementation. As long as one or the other can be driven via cmake (if required), we are OK.  In this case, we will have one and only one implementation, so I am actually in favor of removing this extra layer.

Reply:
The extra layer is required because we need to keep all headers in lib/include/platform platform-independent.

As noted above, the current system is:
```c
typedef pthread_mutex_t _IotSystemMutex_t; // platform specific file in platform/include
typedef _IotSystemMutex_t IotMutex_t; // platform independent file in lib/include
```
So, we can't do
```c
typedef pthread_mutex_t IotMutex_t
```
because it breaks the platform-independent nature of iot_threads.h

Reply:
You are saying tht lib/include should know about _IotSystemMutex_t, this can only be true if the file in lib/include is including the file from platform/include, which it does via the config file.

But this means that if platform/include were to define `typedef pthread_mutex_t IotMutex_t` and if we were to not have _IotSystemMutex_t at all you would have the same properties with one fewer type?

I see no value added by _IotSystemMutex_t here.